File: ts_sdk/src/models/severity.ts
/**
 * Severity tier — single source of truth across the workflow-runner UI.
 *
 * Analyzers have drifted across cycles (we observed at least three vocabularies:
 * `blocker/derived/info`, `error/warn/info`, `high/medium/low`). Every consumer
 * used to reinvent normalization. This tier collapses all of them into three
 * buckets, modeled on Sentry's Issue Priority (High/Medium/Low).
 *
 * UI rules per tier:
 *   ATTENTION — pinned banner + red gutter chip; visible in simple + expert
 *   NOTABLE   — soft yellow gutter chip; visible in simple + expert
 *   INFO      — gray gutter chip; HIDDEN in simple mode, shown only in expert
 */
export enum SeverityTier {
  ATTENTION = 'attention',
  NOTABLE = 'notable',
  INFO = 'info',
}

const ATTENTION_TOKENS = new Set([
  'attention',
  'error',
  'blocker',
  'critical',
  'fatal',
  'high',
  'sut_regression',
  'regression',
  'sla_violation',
  'status_mismatch',
]);

const NOTABLE_TOKENS = new Set([
  'notable',
  'warn',
  'warning',
  'medium',
  'retry',
  'wrong_tool',
  'mid_run_toolsearch',
  'incomplete',
  'protocol_violation',
  'visibility_check',
  'visibility_heuristic_override',
]);

const INFO_TOKENS = new Set([
  'info',
  'low',
  'derived',
  'observation',
  'behavior',
  'latency',
]);

/**
 * Classify an analyzer issue into a tier.
 *
 * Accepts whatever fields the analyzer produced — `severity`, `kind`,
 * `category`, or no signal at all. Order of precedence is severity → kind →
 * category. Unknown tokens default to NOTABLE so the user still sees them
 * in simple mode (better to over-show once than to lose a real issue).
 *
 * @example
 *   classifySeverity('error') === SeverityTier.ATTENTION
 *   classifySeverity('high', 'sut_regression') === SeverityTier.ATTENTION
 *   classifySeverity(undefined, 'wrong_tool') === SeverityTier.NOTABLE
 *   classifySeverity('info') === SeverityTier.INFO
 *   classifySeverity() === SeverityTier.NOTABLE  // unknown ≠ silent
 */
export function classifySeverity(
  rawSeverity?: string | null,
  rawKind?: string | null,
  rawCategory?: string | null,
): SeverityTier {
  for (const raw of [rawSeverity, rawKind, rawCategory]) {
    if (!raw) continue;
    const token = String(raw).trim().toLowerCase();
    if (!token) continue;
    if (ATTENTION_TOKENS.has(token)) return SeverityTier.ATTENTION;
    if (NOTABLE_TOKENS.has(token)) return SeverityTier.NOTABLE;
    if (INFO_TOKENS.has(token)) return SeverityTier.INFO;
  }
  return SeverityTier.NOTABLE;
}

/** Tier ordering for sort. Higher number = more urgent. */
export const SEVERITY_RANK: Record<SeverityTier, number> = {
  [SeverityTier.ATTENTION]: 2,
  [SeverityTier.NOTABLE]: 1,
  [SeverityTier.INFO]: 0,
};

/** Whether a tier is visible in simple mode. */
export function isVisibleInSimpleMode(tier: SeverityTier): boolean {
  return tier !== SeverityTier.INFO;
}

File: flow_sdk/transcript_analyzer/severity.py
"""Severity classifier — Python mirror of ``ts_sdk/src/models/severity.ts``.

Single source of truth for collapsing analyzer-emitted severity vocabularies
(``error/warn/info``, ``high/medium/low``, ``blocker/derived/info``) into
three canonical tiers. Used by future analyzer skill builds and any
backend cost/quality dashboards.

Keep the token sets in sync with the TypeScript file.
"""

from __future__ import annotations

from enum import Enum
from typing import Optional


class SeverityTier(str, Enum):
    """Three-tier severity model — see ``severity.ts`` for UI rules."""

    ATTENTION = "attention"
    NOTABLE = "notable"
    INFO = "info"


_ATTENTION_TOKENS = frozenset(
    {
        "attention",
        "error",
        "blocker",
        "critical",
        "fatal",
        "high",
        "sut_regression",
        "regression",
        "sla_violation",
        "status_mismatch",
    }
)

_NOTABLE_TOKENS = frozenset(
    {
        "notable",
        "warn",
        "warning",
        "medium",
        "retry",
        "wrong_tool",
        "mid_run_toolsearch",
        "incomplete",
        "protocol_violation",
        "visibility_check",
        "visibility_heuristic_override",
    }
)

_INFO_TOKENS = frozenset(
    {
        "info",
        "low",
        "derived",
        "observation",
        "behavior",
        "latency",
    }
)


def classify_severity(
    raw_severity: Optional[str] = None,
    raw_kind: Optional[str] = None,
    raw_category: Optional[str] = None,
) -> SeverityTier:
    """Collapse analyzer-emitted severity strings to a canonical tier.

    Precedence: ``severity → kind → category``. Unknown tokens default to
    NOTABLE so the user still sees them (better to over-show than lose a
    real issue).

    Mirrors ``classifySeverity`` in ``ts_sdk/src/models/severity.ts``.
    """
    for raw in (raw_severity, raw_kind, raw_category):
        if not raw:
            continue
        token = str(raw).strip().lower()
        if not token:
            continue
        if token in _ATTENTION_TOKENS:
            return SeverityTier.ATTENTION
        if token in _NOTABLE_TOKENS:
            return SeverityTier.NOTABLE
        if token in _INFO_TOKENS:
            return SeverityTier.INFO
    return SeverityTier.NOTABLE


SEVERITY_RANK: dict[SeverityTier, int] = {
    SeverityTier.ATTENTION: 2,
    SeverityTier.NOTABLE: 1,
    SeverityTier.INFO: 0,
}


def is_visible_in_simple_mode(tier: SeverityTier) -> bool:
    return tier is not SeverityTier.INFO
